src/conduits/RwaInputConduit3.sol

```diff
@@ -0,0 +1,183 @@
+// Copyright (C) 2020, 2021 Lev Livnev <lev@liv.nev.org.uk>
+// Copyright (C) 2022 Dai Foundation
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+// SPDX-License-Identifier: GPL-3.0-or-later
+pragma solidity 0.6.12;
+
+import {DSTokenAbstract} from "dss-interfaces/dapp/DSTokenAbstract.sol";
+import {PsmAbstract} from "dss-interfaces/dss/PsmAbstract.sol";
+import {GemJoinAbstract} from "dss-interfaces/dss/GemJoinAbstract.sol";
+
+/**
+ * @author Lev Livnev <lev@liv.nev.org.uk>
+ * @author Kaue Cano <kaue@clio.finance>
+ * @author Henrique Barcelos <henrique@clio.finance>
+ * @author Nazar Duchak <nazar@clio.finance>
+ * @title An Input Conduit for real-world assets (RWA).
+ * @dev This contract differs from the original [RwaInputConduit](https://github.com/makerdao/MIP21-RWA-Example/blob/fce06885ff89d10bf630710d4f6089c5bba94b4d/src/RwaConduit.sol#L20-L39):
+ *  - The caller of `push()` is not required to hold MakerDAO governance tokens.
+ *  - The `push()` method is permissioned.
+ *  - `push()` permissions are managed by `mate()`/`hate()` methods.
+ *  - Require PSM and corresponding GEM addresses in constructor
+ *  - The `push()` method swaps GEM to DAI using PSM
+ */
+contract RwaInputConduit3 {
+    /// @dev This is declared here so the storage layout lines up with RwaInputConduit.
+    DSTokenAbstract private __unused_gov;
+    /// @notice Dai token contract address
+    DSTokenAbstract public dai;
+    /// @notice RWA urn contract address
+    address public to;
+
+    /// @notice Addresses with admin access on this contract. `wards[usr]`
+    mapping(address => uint256) public wards;
+    /// @notice Addresses with push access on this contract. `may[usr]`
+    mapping(address => uint256) public may;
+
+    /// @notice GEM token contract address (should ERC20 compliant)
+    DSTokenAbstract public gem;
+    /// @notice PSM contract address for the GEM
+    PsmAbstract public psm;
```

Comment:
As i remember `immutable` takes storage slot as well.

Reply:
From the [docs](https://docs.soliditylang.org/en/v0.8.16/contracts.html?highlight=immutable#immutable):

> Constant and Immutable State Variables[](https://docs.soliditylang.org/en/v0.8.16/contracts.html?highlight=immutable#constant-and-immutable-state-variables)
>
> State variables can be declared as constant or immutable. In both cases, the variables cannot be modified after the contract has been constructed. For constant variables, the value has to be fixed at compile-time, while for immutable, it can still be assigned at construction time.
>
> It is also possible to define constant variables at the file level.
>
> **The compiler does not reserve a storage slot for these variables, and every occurrence is replaced by the respective value.**

Reply:
I remember at one of the debugging spell session we find out out that `immutable` actually reserve slot

Reply:
Not sure maybe I am wrong